lib/ash_authentication/strategies/api_key/sign_in_preparation.ex

```diff
@@ -0,0 +1,114 @@
+defmodule AshAuthentication.Strategy.ApiKey.SignInPreparation do
+  @moduledoc """
+  Prepare a query for sign in.
+  """
+
+  use Ash.Resource.Preparation
+  alias AshAuthentication.{Errors.AuthenticationFailed, Info}
+  alias Ash.{Query, Resource.Preparation}
+  require Ash.Query
+
+  alias Ash.Error.Framework.AssumptionFailed
+
+  @doc false
+  @impl true
+  @spec prepare(Query.t(), keyword, Preparation.Context.t()) :: Query.t()
+  def prepare(query, _opts, _context) do
+    with {:ok, strategy} <- Info.strategy_for_action(query.resource, query.action.name),
+         {:ok, api_key} <- Query.fetch_argument(query, :api_key),
+         {:ok, api_key_id, random_bytes} <- decode_api_key(api_key) do
+      query
+      |> Query.set_context(%{private: %{ash_authentication?: true}})
+      |> Ash.Query.filter(
+        exists(
+          ^[strategy.api_key_relationship],
+          id == ^api_key_id
+        )
+      )
+      |> Query.after_action(fn
+        _query, [record] ->
+          verify_hash_and_set_api_key(record, query, api_key_id, random_bytes, strategy, api_key)
+
+        _query, [] ->
+          Plug.Crypto.secure_compare(
+            :crypto.hash(:sha256, random_bytes <> api_key_id),
+            Ecto.UUID.bingenerate() <> :crypto.strong_rand_bytes(32)
+          )
```

Comment:
The timing is off for this branch since there's no second `Ash.load`.
Is there a way to do both queries in one to avoid the difference?

Reply:
We don't actually have a great way to load associations in a single query in Ash, we've avoided it as (except for this one scenario now) its never really been anything other than an optimization. There are things that can be done using calculations, but that is fairly difficult to make generic in an extension like this... 🤔 I don't want to make a fake query to the database with a random UUID either though 😓.

🤔